README
        Idel: a virtual machine for mobile code
        http://wry.me/~darius/software/idel/
        Version 0.1.10

        By Darius Bacon <darius@wry.me>
        Distributed under the GNU Lesser General Public License; 
        see the accompanying COPYING file for more details.


QUICK START

To build:               ./configure && make
To test:                make test
To run:                 ./run tests/hello.idel
To install:             make install
To remove:              make uninstall
To start a jobber:      bin/jobber (executed from this directory)
To stop it:             bin/kill-jobber


ABOUT

Idel is a virtual machine for safely and portably running untrusted
code.  This is not a new programming language, but a substrate for
existing ones.  It's designed to:

 * be relatively language neutral
 * control resource consumption
 * support mobility and persistence
 * need only an auditably small trusted computing base
 * yield small object files and fast execution
 * accommodate capability-based security in a coming release
 * be easy to incorporate in other programs, the way Tcl is.

See:
 notes/Overview.text    for more motivation.
 doc/tutorial.text      for an intro to Idel's assembly language.
 man idelasm            about invoking the assembler.
 man idelvm             about invoking the runtime.
 examples/borednet      for an example application (CPU sharing)
 examples/roshambo      for another amusing example (bot wars)
 doc/specification.text for an (incomplete) virtual machine spec.

This preview release isn't meant for serious use.  Some important
features are missing, and others will change incompatibly later on.


MANIFEST

 doc/                   Man pages, tutorial, and specification.
 examples/              Some small example programs.
 src/                   The source code.
 tests/                 Test cases.
 benchmarks/            Microbenchmarks.
 contrib/               Contributed code.
 notes/                 Design notes, mostly to myself.


PORTABILITY

To build this, you need:

 * GNU Make
 * awk
 * an ANSI C (C89) compiler

Also, the borednet and roshambo examples use Python, and the test
suite uses the C preprocessor (cpp) that ought to come with your C
compiler but occasionally doesn't.

It's been tested on the following systems:

        CPU             OS               Compiler
        -----------     ------------     ------------
        Alpha           Linux 2.2        gcc 2.95.2
        i386            FreeBSD 4.5      gcc 2.95.3
        i686            Linux 2.0.36     gcc 2.95.4
        i686            Linux 2.4        gcc 2.95.2
        PowerPC G4      Darwin 5.4       gcc 2.95.2
        RS/6000         Linux 2.4        gcc 2.95.2
        Sparc Ultra60   Linux 2.2        gcc 2.95.2
        Sparc R220      Solaris          gcc 2.95.2
        Sparc R220      Solaris          Sun WorkShop 6 update 1 C 5.2 [1]

[1] For Sun's compiler, you must set the -Xa option in CFLAGS to get
it to expect ANSI C.  Changing -O2 to -fast looks like a good idea, too.

Earlier versions were also tested on these, and may still work:

        i586            FreeBSD 4.2      gcc 2.95.2
        i686            Win2k/Cygwin     gcc 2.95.3-5
        PowerPC         Darwin 1.2       gcc 2.7.2.1

If it doesn't work out of the box for you, please let me know.  


ACKNOWLEDGEMENTS

Thanks to:

 * erights.org and Shae Erisson for much-needed donations
 * Luke Gorrie for roshambo and idelrun.el
 * Steve Jenson, Ryan Ziegler, and Glyph Lefkowitz for test accounts


LICENSE

Idel mobile code virtual machine
Copyright (C) 2001-2002  Darius Bacon

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA


ChangeLog
-*- text -*-

New in 0.1.11pre ()

 - tweaked the test scripts to work with newer cpp


New in 0.1.10 (1 August 2002):

 - added flush-output instruction
 - examples/roshambo contributed by Luke Gorrie


New in 0.1.9 (11 May 2002):

 - now works on the Alpha, and should be 64-bit clean in general
 - added doc/cvsinstall.text
 - bugfix: when -profile was on and a basic block got split across
   dynamically allocated code chunks, the profile output would barf
 - bugfix: exit status of idelvm should be 128 when command line args
   are wrong, not 1
 - fixed vm->at_jump_target (it was misplaced, but not in a way that
   currently matters)
 - tweaked test scripts to run on Solaris


New in 0.1.8 (15 Feb 2002):

 - moved executables into bin/
 - fixed makefiles to work with sh as well as bash
 - fixed borednet Python code to conform to coding standards
 - some cleanup of object encoding/decoding
 - added bin/jobber
 - small borednet improvements
 - hashcash demo for borednet


New in 0.1.7 (3 Feb 2002):

bugfixes:
 - assembler was misparsing # line directives from the C preprocessor,
   yielding wrong line numbers in error messages
 - assembler `ints:' directive now word-aligns the new symbol

other:
 - added profiling and -profile switch
 - speedup by combining instruction sequences into superinstructions
 - speedup by specializing instructions to particular immediate
   arguments
 - a couple of other performance tweaks to the interpreter
   (improved scheduling of next-instruction fetch, and offset to stack
   pointer)
 - distributed work of returning() throughout codegen in preparation
   for multiple backends (no more inspection of already-generated
   instructions)
 - added bss-data: declarations to the assembler
 - added character constants to the assembler
 - added SHA-1, quicksort, Boyer-Moore-Horspool, and sieve benchmarks
 - improved codegen error messages
 - renamed `:' to `def' since it's not the same as the Forth `:'
 - renamed the shift operators with the Java names (<<, >>, >>>)


New in 0.1.6 (18 Jan 2002):

bugfixes:
 - assembler occasionally segfaulted on undefined defns

other:
 - assembler supports hex constants
 - assembler is smarter about the Zeroes tag


New in 0.1.5 (16 Jan 2002):

incompatible changes:
 - stack pickling is now disabled unless you run idelvm with the
   -development flag.  Documenting the design revealed flaws in it that
   just aren't worth the trouble right now to fix.  (See notes/stacks.text)

other:
 - added examples/borednet
 - now has an API and builds a library implementing it
 - added Zeroes tag to object file format

	
New in 0.1.4 (18 Dec 2001):

incompatible changes:
 - execution now starts at the first defn, not the last
   (this will enable pipelined downloading and execution of object files)
 - idel programs now must return an exit code

bugfixes:
 - fixed vm bug (stack checker skipped over implicit ELSE clauses)

other:
 - eliminated recursive makefiles (now requires GNU Make)
 - conditionalized gcc-specific flags in Makefile
 - increased assembler hashtable size
 - added byte-data sections to the object format
 - assembler chooses between byte-data and int-data sections intelligently
 - got rid of magic numbers for word size all over the place
 - added huffman example
 - made sure of signed/unsigned mismatches compiler warned about
 - removed hard limit on token length in idelasm
 - minor code cleanup
 - tests check the status code returned, not just the output
 - added dependencies on idelasm, idelvm in Module.mk of some examples


New in 0.1.3 (12 Dec 2001):

incompatible changes:
 - now forbids unaligned word access

bugfixes:
 - fixed assembler bug (allowed reference to out-of-scope local)

other:
 - changed name
 - added calculator example
 - added testexamples target to the makefile
 - integer div, mod in portability.h
 - changed makefile to put idelasm & idelvm in the main directory
 - added a contrib/ directory with Luke Gorrie's elisp code
 - using cpp as preprocessor to the assembler
 - assembler gives source locations in error messages
 - assembler understands line-number directives from cpp
 - assembler may take multiple input files
 - space and time limits are arguments to idelvm instead of constants
 - added some microbenchmarks
 - runtime error messages now come prefixed with the program name
 - added -version, -help options
 - split out the main() from codegen.c
 - man pages
 - added idelrun and cppidelasm scripts for development convenience
 - added an option to set the #! line in idelasm
 - added doc/tutorial.text
 - some code cleanup


src/idelvmmain.c
/*
 * Load and run an Idel VM, from the shell.
 * Copyright (C) 2001-2002 Darius Bacon
 *
 * This is also an example of how to use the API.
 */

#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "idel.h"


/* A fail-stop fopen(), treating a filename of "-" as stdin/stdout. */
static FILE *
open_file (const char *filename, const char *mode)
{
  if (strcmp (filename, "-") == 0) 
    {
      assert (mode[0] == 'r' || mode[0] == 'w');
      return mode[0] == 'w' ? stdout : stdin;
    } 
  else
    {
      FILE *result = fopen (filename, mode);
      if (!result)
	idel_die ("%s: %s", filename, strerror (errno));
      return result;
    }
}


static int
int_value (const char *s)
{
  int n;
  char *endptr;

  errno = 0;
  n = strtol (s, &endptr, 10);
  if (endptr[strspn (endptr, " \t\n")] != '\0' || errno == ERANGE)
    idel_die ("Not a 32-bit integer: %s", s);

  return n;
}


static void
show_version (void)
{
  printf ("idelvm version %s\n", idel_version_string);
  exit (0);
}


static void
show_help (int exit_status)
{
  printf (
"Usage: %s [option]... [object-file]\n", idel_program_name);
  printf (
"where the legal options are:\n"
"  -data n\n"
"  -fuel n\n"
"  -help\n"
"  -stack n\n"
"  -version\n"
"Report bugs to darius@wry.me.\n"
);
  exit (exit_status);
}


const char *idel_program_name;

/* Load and run an object file in a fresh VM. */
int
main (int argc, char **argv)
{
  int stack_size = 200;		/* defaults */
  int data_size = 64 * 1024;
  int fuel = 100000000;
  int profiling = 0;

  idel_OR *or;
  FILE *in = stdin;

  idel_program_name = argv[0];

  /* Parse command-line options */
  for (;;)
    {
      if (1 < argc)
	{
	  if (0 == strcmp ("-version", argv[1]))
	    show_version ();
	  if (0 == strcmp ("-help", argv[1]))
	    show_help (0);
	  if (0 == strcmp ("-development", argv[1]))
	    {
	      idel_development_enabled = 1;
	      --argc, ++argv;
	    }
	  else if (0 == strcmp ("-profile", argv[1]))
	    {
	      profiling = 1;
	      --argc, ++argv;
	    }
	}
      if (2 < argc)
	{
	  if (0 == strcmp ("-stack", argv[1]))
	    stack_size = int_value (argv[2]);
	  else if (0 == strcmp ("-data", argv[1]))
	    data_size = int_value (argv[2]);
	  else if (0 == strcmp ("-fuel", argv[1]))
	    {
	      fuel = int_value (argv[2]);
	      if (fuel < 0)
		idel_die ("Negative fuel: %d", fuel);
	    }  
	  else
	    break;
	  argc -= 2, argv += 2;	  
	}
      else
	break;
    }

  /* Read the object file */
  if (2 < argc)
    show_help (128);
  else if (2 == argc)
    in = open_file (argv[1], "rb");
  or = idel_read_program (in);
  fclose (in);

  /* Run the program */
  {
    int status;
    idel_VM *vm = idel_vm_make (stack_size, data_size, profiling);
    idel_vm_load (vm, or);
    status = idel_vm_run (vm, fuel);
    if (profiling)
      idel_vm_dump_profile (vm, stderr);
    return 0x7f & status;
  }
}
